ddtrace/llmobs/_integrations/langchain.py

```diff
@@ -54,9 +58,71 @@
 SUPPORTED_OPERATIONS = ["llm", "chat", "chain", "embedding", "retrieval", "tool"]
 
 
+def _extract_bound(instance):
+    """
+    Extracts the bound object from a RunnableBinding instance.
+    LangChain will sometimes stick things we trace (chat models) in these bindings,
+    and they will show up as steps in the chain instead of the model instance itself.
+    """
+    if hasattr(instance, "bound"):
+        return instance.bound
+    return instance
+
+
+def _flattened_chain_steps(steps: List[Any], nested: bool = True) -> List[Any]:
+    """
+    Flattens the contents of a chain into non-RunnableBindings and non-RunnableParallel steps.
+    RunnableParrallel steps are extracted and can either be nested into sublists or flattened.
+    """
+    flattened_steps = []
+    for step in steps:
+        step = _extract_bound(step)
+        if hasattr(step, "steps__"):
+            parallel_steps = getattr(step, "steps__", {})
+            flattened_parallel_steps = []
+            for parallel_step in parallel_steps.values():
+                parallel_step = _extract_bound(parallel_step)
+                flattened_parallel_steps.append(parallel_step)
+            if nested:
+                flattened_steps.append(flattened_parallel_steps)
+            else:
+                flattened_steps.extend(flattened_parallel_steps)
+        else:
+            flattened_steps.append(step)
+    return flattened_steps
+
+
 class LangChainIntegration(BaseLLMIntegration):
     _integration_name = "langchain"
 
+    _chain_steps: Set[int] = set()
+    """Set of instance ids that are steps in a chain."""
+
+    _spans: Dict[int, Span] = {}
+    """Maps instance ids to spans."""
+
+    _instances: WeakKeyDictionary = WeakKeyDictionary()
```

Comment:
Just confirming here - is the weak key dict meant to avoid memory leaks / overhead? Do we need to perform any manual clearing of `_spans` or is this sufficient by itself?

Reply:
yeah use it to avoid memory leaks. and yes good callout, i meant to see if we could do some kind of manual clearing for the other two data structures used here, will follow up if i can do that (i think the set should be fine to delete from, but i'm not sure about `_spans`)